### PR TITLE
VirtualGateway API 

### PR DESCRIPTION
**Pull request by ilackarms**

### CODE REVIEW

File: api/common/v1/request_matchers.proto [hunk 1]
```diff
@@ -46,3 +47,47 @@ message StatusCodeMatcher {
...
     }
 }
+
+// Specify HTTP request level match criteria. All specified conditions must be satisfied for a match to occur.
+message HttpMatcher {
+
+    // The name assigned to a match. The match's name will be
+    // concatenated with the parent route's name and will be logged in
+    // the access logs for requests matching this route.
+    string name = 8;
```
Comment (ilackarms) [inline api/common/v1/request_matchers.proto:57]: nit: fix numbering

File: api/common/v1/request_matchers.proto [hunk 2]
```diff
@@ -46,3 +47,47 @@ message StatusCodeMatcher {
...
+
+        // If specified, the targeted path must begin with the prefix.
+        string prefix = 1;
+
+        // If specified, the targeted path must exactly match the value.
+        string exact = 2;
+
+        // If specified, the targeted path must match the regex.
+        string regex = 3;
+    }
```
Comment (ilackarms) [inline api/common/v1/request_matchers.proto:70]: switch `oneof path_specifier` to simply use `StringMatch uri/path`

File: api/common/v1/selectors.proto [hunk 3]
```diff
@@ -118,3 +118,50 @@ message IdentitySelector {
...
+            Exists = 5;
+            // !
+            DoesNotExist = 6;
+            // gt
+            GreaterThan = 7;
+            // lt
+            LessThan = 8;
+        }
+    }
+}
```
Comment (ilackarms) [inline api/common/v1/selectors.proto:167]: move this to skv2

File: api/discovery/v1/mesh.proto [hunk 4]
```diff
@@ -56,6 +56,11 @@ message MeshSpec {
 
         // Describes the ingress gateway.
         message IngressGatewayInfo {
+            // The name of the Ingress Gateway Service
+            string name = 8;
+
+            // The namespace in which the ingress gateway is running.
+            string namespace = 7;
```
Comment (ilackarms) [inline api/discovery/v1/mesh.proto:63]: remove these as ingress gw info now contains a ref

File: api/enterprise/networking/v1beta1/route.proto [hunk 5]
```diff
@@ -0,0 +1,155 @@
+syntax = "proto3";
```
Comment (ilackarms) [inline api/enterprise/networking/v1beta1/route.proto:1]: we should review all the comments and proto number ordering for the new APIs being added to `networking.enterprise.mesh.gloo.solo.io`

File: api/enterprise/networking/v1beta1/route.proto [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+
+    // // TODO: Verify that we need these, and see if there's a cleaner now that we can make breaking changes.
+    // // Whether this route as a child should inherit headers, methods, and query parameter matchers from the parent.
+    // // Defaults to value of parent; for virtual services (no parent) defaults to false.
+    // google.protobuf.BoolValue inheritable_matchers = 3;
+
+    // // Whether this route as a child should inherit path matchers (i.e., path itself, case-sensitive setting) from
+    // // the parent. Defaults to value of parent; for virtual services (no parent) defaults to false.
+    // google.protobuf.BoolValue inheritable_path_matchers = 4;
```
Comment (ilackarms) [inline api/enterprise/networking/v1beta1/route.proto:45]: remove these

File: api/networking/v1/weighed_destination.proto [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        VirtualDestination virtual_destination = 3;
+
+        // Reference to a gloo mesh Static Destination
+        .core.skv2.solo.io.ObjectRef static  = 4;
+
+        // Envoy will determine the cluster to route to by reading the value of the HTTP header named by cluster_header from the request headers.
+        // If the header is not found or the referenced cluster does not exist, Envoy will return a 404 response.
+        // Avoid using this whenever possible, it does not allow for custom filter configuration based on Virtual Host.
+        // {{/* NOTE: unimplemented */}}
+        string cluster_header = 5;
```
Comment (Sodman) [inline api/networking/v1/weighed_destination.proto:35]: cluster header doesn't work with envoy weighted clusters.

File: api/enterprise/networking/v1beta1/virtual_gateway.proto [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+*/
+message VirtualGatewaySpec {
+
+    // deploy this VirtualGateway to one or more Ingress Gateway workloads
+    // {{/* TODO: evaluate supporting multiple ingress gateway deployments per VG */}}
+    DeployToIngressGateway deploy_to_ingress_gateways = 1;
+
+    // deploy this VirtualGateway to one or more workload sidecars
+    // {{/* NOTE: unimplemented */}}
+    repeated .common.mesh.gloo.solo.io.WorkloadSelector deploy_to_sidecars = 2;
```
Comment (Sodman) [inline api/enterprise/networking/v1beta1/virtual_gateway.proto:34]: Hide from docs?

File: api/enterprise/networking/v1beta1/virtual_gateway.proto [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+        // This is required when more than one `connectionHandler` is specified for a single gateway.
+        // Typically this is used to serve different 
+        ConnectionMatch connection_match = 1;
+
+        // The type of routes this handler will serve.
+        // HttpRoutes creates a listener with an http_connection_manager.
+        // TcpRoutes creates a listener with a tcp proxy filter.
+        oneof handler_type {
+            HttpRoutes http = 2;
+            TcpRoutes tcp = 3;
```
Comment (Sodman) [inline api/enterprise/networking/v1beta1/virtual_gateway.proto:80]: Note: TCP unimplemented

File: api/networking/v1/weighed_destination.proto [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    oneof destination_type {
+
+        // Specify a Kubernetes Service.
+        KubeDestination kube_service = 2;
+
+        // Specify a VirtualDestination.
+        VirtualDestination virtual_destination = 3;
+
+        // Reference to a gloo mesh Static Destination
+        .core.skv2.solo.io.ObjectRef static  = 4;
```
Comment (Sodman) [inline api/networking/v1/weighed_destination.proto:29]: rename to static_destinatination

File: api/discovery/v1/mesh.proto (hunk 4 above)
Comment (Sodman) [reply api/discovery/v1/mesh.proto:65]: Remove after we update the translator logic to pull this data from the destination discovery object (which should be referenced in the mesh proto - which still needs to be implemented).

File: api/networking/v1/destination_policy.proto [hunk 11]
```diff
@@ -0,0 +1,102 @@
+syntax = "proto3";
```
Comment (Sodman) [inline api/networking/v1/destination_policy.proto:1]: Need to discuss /w team: Do we need to implement this before merge?

Comment (Sodman) [reply api/networking/v1/destination_policy.proto]: Update: We do not. TODO: Remove destination_policy from this PR and open as a new PR/issue.

As well as outlier detection and mtls config, we will also want to add tcpkeepalive as an option here.

File: api/enterprise/networking/v1beta1/route.proto [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    // Route options include configuration such as retries, rate limiting, and request/response transformation.
+    // RouteOption behavior will be inherited by delegated routes which do not specify their own `options`
+    .networking.mesh.gloo.solo.io.TrafficPolicySpec.Policy options = 7;
+
+    // RouteActions are used to route matched requests to upstreams.
+    message RouteAction {
+        // Defines the destination upstream for routing
+        // Some destinations require additional configuration for the route (e.g. AWS upstreams require a function name
+        // to be specified).
+        repeated .networking.mesh.gloo.solo.io.WeightedDestination destinations = 1;
```
Comment (yuval-k) [inline api/enterprise/networking/v1beta1/route.proto:63]: we should verify that this is equivalent to the existing envoy model (clusetr, cluster_header, cluster_weight)

File: api/networking/v1/traffic_policy.proto [hunk 13]
```diff
@@ -101,11 +66,19 @@ message TrafficPolicySpec {
 
         // Configure [outlier detection](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier) on the selected destinations.
         // Specifying this field requires an empty `source_selector` because it must apply to all traffic.
+        // DEPRECATED: use DestinationPolicy to configure Outlier Detection
```
Comment (Sodman) [inline api/networking/v1/traffic_policy.proto:69]: Probably should remove deprecation notice here and for mtls until we add destination policy API.

**Review by npolshakova — APPROVED** [review]

**Pull request merged**